lib/LWP/UserAgent.pm

```diff
@@ -897,16 +897,20 @@ sub handlers {
 
 sub run_handlers {
     my($self, $phase, $o) = @_;
+
+    # were we pass $_[2] to the callbacks, instead of $o, so that they
+    # can assign to it; e.g. request_prepare is documented to allow
+    # that
     if (defined(wantarray)) {
         for my $h ($self->handlers($phase, $o)) {
-            my $ret = $h->{callback}->($o, $self, $h);
+            my $ret = $h->{callback}->($_[2], $self, $h);
```

Comment:
This is interesting.  I was just looking at this method recently in #271  Does this change the behaviour of the other handlers as well then?

Reply:
It does. The other handlers' documentation says nothing about what happens when you assign to `@_`, and in general I'd expect a Perl programmer to either not touch `@_`, or know that something weird happens when you do.

I see a few ways to resolve the conflict between the documentation and the implementation:

1. change the docs, say that `request_prepare` has to modify the passed-in object
2. apply this change
3. apply a similar change, but have two different `run_handlers` (or an extra parameter) to select if the handlers can assign to `@_`

The main problem with the current situation (and with approach 3): assigning to `@_` in a (for example) `request_prepare` callback will modify the object inside `run_handlers`, the change will be visible to other `request_prepare` callbacks invoked in the same loop, but will disappear once `run_handlers` returns. This is surprising, and hard to debug.